src/kill-parallel.c

```diff
@@ -0,0 +1,261 @@
+  if (ready == 0 && timeleft >= 0) {
+    do {
+      ready = poll(&fd, 1, timeleft);
+    } while (ready == -1 && errno == EINTR);
+  }
+
+  // If some processes are still running after the grace period,
+  // resort to SIGKILL
+  for (int i = 0; i < ps_n; ++i) {
+    if (killed[i]) {
+      continue;
+    }
+
+    ps_handle_t *handle = R_ExternalPtrAddr(VECTOR_ELT(ps, i));
+
+    if (psll__is_running(handle)) {
+      kill(handle->pid, SIGKILL);
+    }
+  }
+
+ cleanup:
+  kill_parallel_pipe = -1;
+
+  if (fds[0] >= 0) close(fds[0]);
+  if (fds[1] >= 0) close(fds[1]);
+
+  kill_parallel_remove_sigchld(&old_action);
+  procmask_set(&old_mask);
+
+  if (err) {
+    r_unwind(cont);
+  }
+
+  UNPROTECT(2);
+  return R_NilValue;
+}
+
+
+// SIGCHLD -------------------------------------------------------------
+
+static
+void kill_parallel_sigchld_callback(int sig, siginfo_t *info, void *ctx) {
+  int saved_errno = errno;
+
+  /* This might be called on another thread, if the main thread blocks
+     the signal temporarily, so we forward it to the main thread.
+     It is OK to call pthread_self() and pthread_kill() in the signal
+     handler, according to POSIX.1-2008 TC1.
+     https://man7.org/linux/man-pages/man7/signal-safety.7.html */
+  if (pthread_self() != ps__main_thread) {
+    pthread_kill(ps__main_thread, SIGCHLD);
+    errno = saved_errno;
+    return;
+  }
+
+  int res = 0;
+  do {
+    res = write(kill_parallel_pipe, "1", 1);
+  } while (res == -1 && errno == EINTR);
+
+  // Notify next handler
+  if (old_sig_handler.sa_handler != SIG_DFL &&
+      old_sig_handler.sa_handler != SIG_IGN &&
+      old_sig_handler.sa_handler != NULL) {
+    if (old_sig_handler.sa_flags | SA_SIGINFO) {
+      old_sig_handler.sa_sigaction(sig, info, NULL);
+    } else {
+      old_sig_handler.sa_handler(sig);
+    }
+  }
+
+  errno = saved_errno;
+}
+
+static
+void kill_parallel_setup_sigchld(struct sigaction *prev) {
+  if (!prev) {
+    PS__STOP_INTERNAL("Must provide memory to save current handler.");
+  }
+
+  sigset_t old_mask;
+  block_sigchld_save(&old_mask);
+
+  memcpy(prev, &old_sig_handler, sizeof(old_sig_handler));
+
+  struct sigaction action = {{ 0 }};
+  action.sa_sigaction = &kill_parallel_sigchld_callback;
+  action.sa_flags = SA_SIGINFO | SA_RESTART | SA_NOCLDSTOP;
+  sigaction(SIGCHLD, &action, &old_sig_handler);
+
+  procmask_set(&old_mask);
+}
+
+static
+void kill_parallel_remove_sigchld(struct sigaction *prev) {
+  if (!prev) {
+    PS__STOP_INTERNAL("Must provide memory to save current handler.");
+  }
+
+  sigset_t old_mask;
+  block_sigchld_save(&old_mask);
+
+  sigaction(SIGCHLD, &old_sig_handler, NULL);
+  memcpy(&old_sig_handler, prev, sizeof(*prev));
+
+  procmask_set(&old_mask);
+}
+
+
+static
+void block_sigchld_save(sigset_t *old) {
+  sigset_t new;
+  sigemptyset(&new);
+  sigaddset(&new, SIGCHLD);
+
+  if (sigprocmask(SIG_BLOCK, &new, old) == -1) {
+    PS__STOP_INTERNAL("Can't set up sigprocmask");
```

Comment:
For unlikely errors it didn't seem worth the complications of returning an error object to try and recover. It seems the most important thing is to report what happens, so I've added some tooling to throw internal errors.